### PR TITLE
Data Service, Course, and Category Unit Tests

### PR DESCRIPTION
**Pull request by joshcosta**
This is WIP but I'd like to get some feedback on how the tests are looking and if I'm going about writing them correctly. Let me know your thoughts.

### CODE REVIEW

File: src/app/services/data.service.ts [hunk 1]
```diff
@@ -188,6 +193,11 @@ export class DataService {
 			return false;
 		}
 
+		if (index >= this.courses.length) {
```
Comment (vikaspotluri123) [inline src/app/services/data.service.ts]: How would this happen?

File: src/test/model-classes/category.spec.ts [hunk 2]
```diff
@@ -29,6 +29,50 @@ describe('Unit > Category Model', function () {
...
+			expect(result.valid).to.be.true;
+			expect(result.message).to.equal(null);
+		});
+
+		it('Updating grade to empty string', function () {
+			const result = model.updateGradeValue(0, '');
+
+			expect(result.valid).to.be.true;
+			expect(result.message).to.equal(null);
+		});
```
Comment (vikaspotluri123) [inline src/test/model-classes/category.spec.ts]: I think these can be merged (e.g. `it('falsy values')`)

File: src/test/model-classes/category.spec.ts [hunk 3]
```diff
@@ -29,6 +29,50 @@ describe('Unit > Category Model', function () {
...
+
+		it('Updating with valid grade', function () {
```
Comment (vikaspotluri123) [inline src/test/model-classes/category.spec.ts]: Looks like the title got flipped with the other one 😄

File: src/test/services/data.service.spec.ts [hunk 4]
```diff
@@ -1,10 +1,294 @@
...
+
+describe('Unit > Data Service', () => {
+	const service = new DataService(
+		new FakeAPIService() as unknown as ApiService,
+		{} as UserService,
+		new FakeNotifyService() as unknown as NotificationService
+	);
+	const api: ApiService = new FakeAPIService() as unknown as ApiService;
+
+	describe('reset', () => {
```
Comment (vikaspotluri123) [inline src/test/services/data.service.spec.ts]: Unless you're planning on adding more reset tests, you can just make this an `it` block

File: src/test/services/data.service.spec.ts [hunk 5]
```diff
@@ -1,10 +1,294 @@
...
+	}
+}
+
+class FakeNotifyService {
+	notify(..._args): {} {
+		return {};
+	}
+}
+
+describe('Unit > Data Service', () => {
```
Comment (vikaspotluri123) [inline src/test/services/data.service.spec.ts]: Don't use anonymous functions in tests

File: src/test/services/data.service.spec.ts [hunk 6]
```diff
@@ -1,10 +1,294 @@
...
+			service.setCurrentSemester('2100F');
+			expect(service.currentSemester.raw).to.equal('2100F');
+		});
+
+		it('Updating with invalid year', () => {
+			service.setCurrentSemester('10F');
+			expect(service.currentSemester.raw).to.be.undefined;
+		});
+
+		it('Updating with invalid letter', () => {
```
Comment (vikaspotluri123) [inline src/test/services/data.service.spec.ts]: Maybe "season" instead of "letter"... it doesn't really matter though

File: yarn.lock [hunk 7]
```diff
@@ -5959,6 +5959,11 @@ lolex@^4.1.0, lolex@^4.2.0:
   resolved "https://registry.yarnpkg.com/lolex/-/lolex-4.2.0.tgz#ddbd7f6213ca1ea5826901ab1222b65d714b3cd7"
   integrity sha512-gKO5uExCXvSm6zbF562EvM+rd1kQDnB9AZBbiQVzf1ZmdDpxUSvpnAaVOP83N/31mRK8Ml8/VE8DMvsAZQ+7wg==
 
+lolex@^4.1.0, lolex@^4.2.0:
```
Comment (vikaspotluri123) [inline yarn.lock]: I'm not sure that this should have changed 😆

File: src/app/services/data.service.ts (hunk 1 above)
Comment (joshcosta) [reply src/app/services/data.service.ts]: ehhhh it wouldn't

**Comment by joshcosta** [comment]
Update on this PR: I added unit tests for courses. It even includes a regression test that currently fails because of #58 bug. 

Once @vikaspotluri123 approves I think we should go ahead and merge.

**Pull request merged**